Addressbook::addTag() is not used #169

PierceAndy opened the pull request:
Fixes #169 

Removed unused tag level methods in AddressBook: 
- addTag()
- removeTag()
- containsTag()

@yamgent @chao1995 With your unit test experiences, what are your thoughts on how tests that relied on containsTag() were handled?

CanIHasReview commented:
# v1

@PierceAndy submitted v1 for review.

* [v1 1/3] [AddressBook: remove unused addTag\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/719359d08d71293abe1e7c230023ce49e0723601)
* [v1 2/3] [AddressBook: remove unused removeTag\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/3ec8d3dec6a439a934fb3c9aef9d2fb1db031d52)
* [v1 3/3] [AddressBook: remove unused containsTag\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/033e8befd3d5effe3d50d9e6753cb467878b37cd)

([:books: Archive](https://github.com/se-edu/addressbook-level2/compare/5810f7c47b3f5392b75184bf95a896df33993744...033e8befd3d5effe3d50d9e6753cb467878b37cd))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level2.git refs/pr/179/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

yamgent commented on src/seedu/addressbook/data/tag/UniqueTagList.java:
```diff
@@ -94,18 +94,6 @@ public boolean contains(Tag toCheck) {
-     *
-     * @throws DuplicateTagException if the Tag to add is a duplicate of an existing Tag in the list.
-     */
-    public void add(Tag toAdd) throws DuplicateTagException {
```
Nice spot, but this should probably be in a separate commit because [v1 1/3] is about `AddressBook#addTag()`, not `UniqueTagList#add()`.

Also, not sure whether it is a good reason to remove `UniqueTagList#add()` just because `AddressBook#addTag()` no longer exist, since I can understand that you are not supposed to manipulate `AddressBook`'s tags directly, but I don't expect `UniqueTagList` to no longer allow adding new tags as a result. @chao1995 Your thoughts on this?

PierceAndy replied:
`AddressBook#addTag()` is the only method that uses `UniqueTagList#add()`, and similarly so with remove in [v1 2/3]. Of course, I can understand the arguments for having a separate issue + PR for unused methods in UniqueTagList.

yamgent replied:
Actually I meant having a separate commit in this current PR, not a separate PR.

chao1995 replied:
> this should probably be in a separate commit

Agree.

> whether it is a good reason to remove UniqueTagList#add() just because AddressBook#addTag() no longer exist

Since UniqueTagList#add() is never used as well. I would suggest we remove it.

yamgent commented on src/seedu/addressbook/data/tag/UniqueTagList.java:
```diff
@@ -118,18 +112,6 @@ public void mergeFrom(UniqueTagList tags) {
-     *
-     * @throws TagNotFoundException if no such Tag could be found in the list.
-     */
-    public void remove(Tag toRemove) throws TagNotFoundException {
```
Same concern as [v1 1/3].

yamgent commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -93,11 +91,11 @@ public void addPerson_emptyAddressBook() throws Exception {
-        assertFalse(defaultAddressBook.containsTag(tagEconomist));
-        assertFalse(defaultAddressBook.containsTag(tagPrizeWinner));
+        assertFalse(defaultAddressBook.getAllTags().contains(tagEconomist));
+        assertFalse(defaultAddressBook.getAllTags().contains(tagPrizeWinner));
```
@PierceAndy I assume that you are talking about this one in the OP of the PR. 

Since `containsTag()` is only used in testing, and we are removing it, this alternative looks acceptable. However, it seems that we also have a method `AddressBookTest#isTagObjectInAddressBookList()` that also allow us to check the tags in the AddressBook (like how it used in line 100). Maybe consider using that instead?

PierceAndy replied:
Yes, it does make sense to use that and be consistent.

PierceAndy commented:
@yamgent About the use of semi-colons, yes it's possible to use semi-colons in place of commas as a kind of "super-comma" when your list is complex. As to whether it's good practice, I guess I might reconsider it.

CanIHasReview commented:
# v2

@PierceAndy submitted v2 for review.

* [v2 1/5] [AddressBook: remove unused addTag\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/5efd64a858424b42ff7aebcbec24664e504da5ee)
* [v2 2/5] [UniqueTagList: remove unused add\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/ef42b3b5fa2081925a957f4351eadfc9ff147e00)
* [v2 3/5] [AddressBook: remove unused removeTag\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/825db626e4696b9c3282efd4d8117667da6e032f)
* [v2 4/5] [UniqueTagList: remove unused remove\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/ac90517510431172abb53014a00216af54b17e14)
* [v2 5/5] [AddressBook: remove unused containsTag\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/ecb4d744648c35a46eb7e163a8613ba56a4901b3)

([:books: Archive](https://github.com/se-edu/addressbook-level2/compare/5810f7c47b3f5392b75184bf95a896df33993744...ecb4d744648c35a46eb7e163a8613ba56a4901b3)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level2/compare/081a810df166cc8249f7daa420b7a7603b13d95e~1...081a810df166cc8249f7daa420b7a7603b13d95e))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level2.git refs/pr/179/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

PierceAndy commented:
# v2

- Split `UniqueTagList` changes into their own commits
- Replaced uses of `AddressBook#getAllTags()` + `UniqueTagList#contains()` in `AddressBookTest` with `AddressBookTest#isTagObjectInAddressBookList()`
- Reworded commit messages based on feedback.

CanIHasReview commented:
# v3

@PierceAndy submitted v3 for review.

* [v3 1/5] [AddressBook: remove unused addTag\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/ef71224cb586bb176a420ea3e364e18092fd9fb8)
* [v3 2/5] [UniqueTagList: remove unused add\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/6bb777c5d0dd438eadda37c7c9e8e1a4beeda42a)
* [v3 3/5] [AddressBook: remove unused removeTag\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/fd6b775a2251b88db5ef98352e88602c8d089c89)
* [v3 4/5] [UniqueTagList: remove unused remove\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/aa541709abf0a0680e54a507b89630c9a561da99)
* [v3 5/5] [AddressBook: remove unused containsTag\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/a5ed6635527717f0be63b3b2bb0c74627e756838)

([:books: Archive](https://github.com/se-edu/addressbook-level2/compare/5810f7c47b3f5392b75184bf95a896df33993744...a5ed6635527717f0be63b3b2bb0c74627e756838)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level2/compare/dc95b862a304de5e9431e72cf534adf48878802b~1...dc95b862a304de5e9431e72cf534adf48878802b))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level2.git refs/pr/179/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

PierceAndy commented:
# v3

- Updated commit messages to reflect updated standards for commit messages here: https://github.com/oss-generic/process/pull/60

PierceAndy commented:
# v3

- Updated commit messages to reflect updated standards for commit messages here: https://github.com/oss-generic/process/pull/60

@yamgent @chao1995

PierceAndy commented:
This PR has two approved peer reviews. Proceeding with next round of reviews with senior dev.
Attn: @damithc

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -93,12 +91,9 @@ public void addPerson_emptyAddressBook() throws Exception {
         defaultAddressBook.addPerson(davidElliot);
-        assertTrue(defaultAddressBook.containsTag(tagEconomist));
-        assertTrue(defaultAddressBook.containsTag(tagPrizeWinner));
-
```
Noticed something odd here. The test does not confirm that the person is added correctly.

damithc commented:
@pyokagan 

> why is AddressBookTest#isTagObjectInAddressBookList() chosen over
> AddressBook#containsTag()?

If we choose `AddressBook#containsTag()` we can only justify its existences with the reasons,
1. For possible future use (but it violates YAGNI)
2. For the use of test cases (but we should not bloat production code just to facilitate testing if there is an alternative) 

Neither reason is strong enough to justify an additional method in a production API? But we can make an exception here arguing that the two reasons combined is enough justification. The method does not complicate the production code that much and _does_ reduce the complexity of test code, so there is no overall additional cost.

pyokagan commented:
Just a side note: I just realised that `AddressBookTest#isTagObjectInAddressBookList()` isn't actually a 1-to-1 replacement of `AddressBook#containsTag()`. The former checks for equal references while the latter checks for equal values.

damithc commented:
> Just a side note: I just realised that AddressBookTest#isTagObjectInAddressBookList() isn't actually a 1-to-1 replacement of AddressBook#containsTag(). The former checks for equal references while the latter checks for equal values.

Oh yes. That was why the method was added in the first place (to confirm the `Person` object points to `Tag` objects in the common tag list instead of keeping it's own copies of the tags). Probably need better method name and better header comment to emphasize this 'special' nature of the method. That also means the method is specific to testing and it has to exist with or without `AddressBook#containsTag()`.

pyokagan commented:
@damithc 
> That also means the method is specific to testing and it has to exist with or without AddressBook#containsTag().

Yes, agreed. In light of this new discovery (which should be emphasized in the commit message), I agree that `AddressBookTest#isTagObjectInAddressBookList()` should stay inside the test code, and that we can remove `AddressBook#containsTag()`. (The alternative would be to make `AddressBook#containsTag()` compare by reference to fully take over the responsibilities of `AddressBookTest#isTagObjectInAddressBookList()`, but that would be hugely inconsistent with our API design.)

CanIHasReview commented:
# v4

@PierceAndy submitted v4 for review.

* [v4 1/5] [AddressBook: remove unused addTag\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/419ec2b8b97a09e5cc9453346669011365b6a999)
* [v4 2/5] [UniqueTagList: remove unused add\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/84c8535cc6562b33a9f70d4a11d2bc9049ac3b7d)
* [v4 3/5] [AddressBook: remove unused removeTag\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/fb8863dfb240d9a1e78734fa3b9fcde21bd2c3c6)
* [v4 4/5] [UniqueTagList: remove unused remove\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/4e3fac92d280d366f638ae21d8e670344fa18891)
* [v4 5/5] [AddressBook: remove unused containsTag\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/4cfe9a5842be32171470115ba50ba0bba77e7c3d)

([:books: Archive](https://github.com/se-edu/addressbook-level2/compare/7492790ac1bb807c443ce4cf24f059fb14fd5442...4cfe9a5842be32171470115ba50ba0bba77e7c3d)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level2/compare/73dea0f6bf7f920bd9df503299004ce9cf429ff6~1...73dea0f6bf7f920bd9df503299004ce9cf429ff6))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level2.git refs/pr/179/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

PierceAndy commented:
# v4

- Created issue for missing verification in AddressBookTest #182 
- Updated [v3 5/5] commit message
> AddressBook: remove unused containsTag()
> 
> AddressBook#containsTag() checks if the master list of Tags contains a
> Tag with the same value as the given Tag.
> 
> This method is not used anywhere, and can be classified as unused code.
> There are no commands that support it, and it is also not part of any
> learning objectives. While it is used in tests, its functionality is
> superseded by AddressBookTest#isTagObjectInAddressBookList(), and can be
> replaced by the latter in these tests. AddressBook#containsTag() should
> be removed as unused code contributes to maintenance and comprehension
> overheads.
> 
> Let's remove AddressBook#containsTag().
> 
> ddressBookTest#isTagObjectInAddressBookList() is chosen over
> AddressBook#containsTag() because the former supersedes the latter in
> functionalities, and the former has to exist with or without the latter
> as the former contains test-specific checks.
> AddressBookTest#isTagObjectInAddressBookList() checks if the master list
> of Tags from a given AddressBook contains the given Tag object by
> reference, thus not only checking if a Tag value is in the master list
> like AddressBook#containsTag() does, but also checks that Person objects
> points to Tag objects in the master Tag list instead of keeping its own
> copies of Tags. The alternative of updating AddressBook#containsTag() to
> compare by reference to take over the responsibilities of
> AddressBookTest#isTagObjectInAddressBookList() is inconsistent with our
> API design, as it would mean adding a method used only in tests to
> production code.

damithc commented:
Tried to shorten the commit message. Does it cover all the points?

```
AddressBook#containsTag() is not used in the production code.
While it is used in test code, its use can be replaced by 
AddressBookTest#isTagObjectInAddressBookList().

Let's remove AddressBook#containsTag().

Notes: 
* Why not keep AddressBook#containsTag() and remove 
AddressBookTest#isTagObjectInAddressBookList() instead?
The former does a value equality test while the latter goes further
and does a reference equality test. The test code requires a 
reference equality test to verify Person objects refer to Tag
objects in the common tag list instead of keeping its own copies
of Tag objects. 
* Why not change AddressBook#containsTag() to use reference
equality and use that in tests?
Doing so will make AddressBook#containsTag()'s semantics 
inconsistent with the rest of our API.
```

CanIHasReview commented:
# v5

@PierceAndy submitted v5 for review.

* [v5 1/5] [AddressBook: remove unused addTag\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/419ec2b8b97a09e5cc9453346669011365b6a999)
* [v5 2/5] [UniqueTagList: remove unused add\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/84c8535cc6562b33a9f70d4a11d2bc9049ac3b7d)
* [v5 3/5] [AddressBook: remove unused removeTag\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/fb8863dfb240d9a1e78734fa3b9fcde21bd2c3c6)
* [v5 4/5] [UniqueTagList: remove unused remove\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/4e3fac92d280d366f638ae21d8e670344fa18891)
* [v5 5/5] [AddressBook: remove unused containsTag\(\)](https://github.com/se-edu/addressbook-level2/pull/179/commits/a6cc952792687098c1790b91936270be6b05ba11)

([:books: Archive](https://github.com/se-edu/addressbook-level2/compare/7492790ac1bb807c443ce4cf24f059fb14fd5442...a6cc952792687098c1790b91936270be6b05ba11)) ([:chart_with_upwards_trend: Interdiff between v4 and v5](https://github.com/se-edu/addressbook-level2/compare/79c993031b6cd755e86d41797bf5d0a68c0a3e79~1...79c993031b6cd755e86d41797bf5d0a68c0a3e79))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level2.git refs/pr/179/5/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

PierceAndy commented:
# v5
- Shortened [v3 5/5] commit message
